### PR TITLE
lightly dataset.subset from filenames me

### LINKED ISSUE #175: Add subset from filenames functionality to LightlyDataset
The `ActiveLearningAgent` returns a list of filenames for which the user will need to get predictions. Since this can be difficult and it's important, that the order of the filenames stays the same, it would be a good option to add the following functionality to the `LightlyDataset`.

```python
dataset = LightlyDataset.subset_from_filenames(input_dir, agent.unlabeled_set)
dataloader = torch.utils.data.DataLoader(dataset, shuffle=False, batch_size=1)
```

## Question A: Which kind of base dataset are supported during the creation?

1. input_dir: `str`
2. dataset: `torchvision.datasets.ImageFolder`
3. dataset: `lightly.data._helpers.DatasetFolder`
4. dataset: `lightly.data._video.VideoDataset`
5. dataset: `generic torchvision dataset` - Problem: it does not necessarily contain filenames
6. dataset: `LightlyDataset`

## Question B: How is the `SubsetDataset` represented:

1. just like the base dataset - not compatible with Option A1
2. a `LightlyDataset`
3. a `LightlyDatasetSubset`

## Question C: How is the `SubsetDataset` created:

1. Add a method to the base dataset - only compatible with option A6
2. Add a constructor to its representation - only compatible with B2 and B3

### PR DESCRIPTION
**Pull request by MalteEbner**
close #175

### CODE REVIEW

File: lightly/data/lighty_subset.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+from lightly.data.dataset import LightlyDataset
+
+
+class LightlySubset(LightlyDataset):
+    def __init__(self, base_dataset: LightlyDataset, filenames_subset: List[str]):
+        self.base_dataset = base_dataset
+
+        dict_base_dataset_filename_index: Dict[str, int] = \
+            dict([
```
Comment (philippmwirth) [inline lightly/data/lighty_subset.py:11]: I think we should refrain from using list comprehension when the expressions are more complex. Why not simply use a loop?

File: lightly/data/lighty_subset.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+                (base_dataset.index_to_filename(self.dataset, index), index)
+                for index in range(len(base_dataset))
+            ])
+        self.mapping_subset_index_baseset = \
```
Comment (philippmwirth) [inline lightly/data/lighty_subset.py:15]: I'd rename it to be more clear: `subset_index_to_baseset_index`, `subset_index_to_dataset_index`, `map_subset_index_to_baseset_index` are possible candidates.

File: lightly/data/lighty_subset.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            [dict_base_dataset_filename_index[filename] for filename in filenames_subset]
+
+    def __getitem__(self, index_subset: int) -> Tuple[object, object, str]:
+        index_baseset = self.mapping_subset_index_baseset[index_subset]
+        sample, target, fname = self.base_dataset.__getitem__(index_baseset)
```
Comment (philippmwirth) [inline lightly/data/lighty_subset.py:20]: You could also leave away this extra line but that's just a minor thing.

File: lightly/data/lighty_subset.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        return sample, target, fname
+
+    def __len__(self) -> int:
+        return len(self.mapping_subset_index_baseset)
+
+    def index_to_filename(self, dataset, index_subset: int):
```
Comment (philippmwirth) [inline lightly/data/lighty_subset.py:26]: Why not simply store the list of filenames which are passed in the beginning and the return the filename at index?

**Review by philippmwirth — COMMENTED** [review]
Logic is straight-forward and looks good. I have some suggestions for better names, please take a look.

An additional problem I see is the fact that the `LightlySubset` is inheriting from and wrapping around `LightlyDataset` at the same time. This looks like it could spell trouble in the future -> very easy source for mistakes and bugs. Don't you think a PyTorch like implementation would be more straight-forward and make more sense although it provides a little less user comfort?

File: lightly/data/lighty_subset.py (hunk 1 above)
Comment (MalteEbner) [reply lightly/data/lighty_subset.py]: I assumed that the list comprehension would be faster. However, now that I profiled it, it isn't, thus I changed it to use a for loop.

File: lightly/data/lighty_subset.py (hunk 3 above)
Comment (MalteEbner) [reply lightly/data/lighty_subset.py:21]: I prefer naming the return values explicitly, as it makes it much easier and faster during coding to assign the return values of a call of this function to variables:
```python
for i in range(len(dataset)):
    sample, target, fname = dataset.__getitem__(i)
```
Without naming the variables explicitly, I would need to dig into the nested function `self.base_dataset.__getitem__(index_baseset)` and possibly even further to get what the return values are and stand for.

**Comment by MalteEbner** [comment]
About the idea to use a pytorch subset instead of a LightlySubset being both a child class and a wrapper:

The current implementation is very close to the one of the [`pytorch Subset`](https://pytorch.org/docs/stable/_modules/torch/utils/data/dataset.html#Subset), which is also both a child and a wrapper. Thus the `pytorch Subset` has the same problems, e.g. its transform is not settable directly as `dataset.transform`, but only nested through `dataset.dataset.transform`. 
Thus using a pytorch Subset is also a potential source for trouble in the future. I would consider it as even more dangerous, as the wrappers for the properties are not made explicitly. 
Another disadvantage of the `pytorch Subset` is that it would not support the `get_filenames` and `dump` methods.

**Comment by philippmwirth** [comment]
> 
> 
> About the idea to use a pytorch subset instead of a LightlySubset being both a child class and a wrapper:
> 
> The current implementation is very close to the one of the [`pytorch Subset`](https://pytorch.org/docs/stable/_modules/torch/utils/data/dataset.html#Subset), which is also both a child and a wrapper. Thus the `pytorch Subset` has the same problems, e.g. its transform is not settable directly as `dataset.transform`, but only nested through `dataset.dataset.transform`.
> Thus using a pytorch Subset is also a potential source for trouble in the future. I would consider it as even more dangerous, as the wrappers for the properties are not made explicitly.
> Another disadvantage of the `pytorch Subset` is that it would not support the `get_filenames` and `dump` methods.

Hm good point, I didn't know that.

**Review by philippmwirth — APPROVED** [review]

**Review by philippmwirth — APPROVED** [review]

**Pull request merged**